### ISSUE TITLE
Kable Hangs IOS app when device disconnects multiple times

### ISSUE THREAD

**Issue by burnhamd** [JuulLabs/kable#308]
Occasionally while using bluetooth background mode I see that when I return to the foreground, if an event occurred in the background that the app is hung.  Looking at the trace, I see that it is almost always hung up in the following code:
dispatch_sync(dispatchQueue) {
            block.run()
        }

I can recreate this issue by disconnecting a device multiple times in the background and seeing that it is called from the cancelPeripheralConnection() in connectAsync()

**Comment by burnhamd** [comment]
Looking more into this, it is not limited to background.
What I have found is that it fails if there is some action taking place like write, or discovery of services.  

It seems that this is being executed on the ApplePeripheral.connection context.  But the actual write gets executed on the central manager dispatcher context.  The problem seems to be when the connection has been closed but the central manager function is running.  Then it just hangs.  Haven't found a fix.  Im not familiar enough with this code to be able to correct it at this time.
